Objects/codeobject.c

```diff
@@ -1925,12 +1925,100 @@ _PyCode_SetUnboundVarCounts(PyThreadState *tstate,
 }
 
 
+const char *
+_PyCode_CheckNoInternalState(PyCodeObject *co)
+{
+    if (_PyCode_HAS_EXECUTORS(co) || _PyCode_HAS_INSTRUMENTATION(co)) {
+        return "only basic code objects are supported";
+    }
+    if (co->_co_monitoring != NULL) {
+        return "only basic code objects are supported";
+    }
+    if (co->co_extra != NULL) {
+        return "only basic code objects are supported";
+    }
+    return NULL;
+}
+
+const char *
+_PyCode_CheckNoExternalState(PyCodeObject *co, _PyCode_var_counts_t *counts)
+{
+    if (counts->numfree > 0) {  // It's a closure.
+        return "closures not supported";
+    }
+    assert(counts->locals.hidden.total == 0);
+
+    if (counts->unbound.globals.numglobal > 0) {
+        return "globals not supported";
+    }
+    else if (counts->unbound.globals.numbuiltin > 0
+             && counts->unbound.globals.numunknown > 0) {
+        return "globals not supported";
+    }
+    // Otherwise we don't check counts.unbound.globals.numunknown since we can't
+    // distinguish beween globals and builtins here.
+
+    return NULL;
+}
+
+int
+_PyCode_VerifyStateless(PyThreadState *tstate,
+                        PyCodeObject *co, PyObject *globalnames,
+                        PyObject *globalsns, PyObject *builtinsns)
+{
+    const char *errmsg;
+   _PyCode_var_counts_t counts = {0};
+    _PyCode_GetVarCounts(co, &counts);
+    if (_PyCode_SetUnboundVarCounts(
+                            tstate, co, &counts, globalnames, NULL,
+                            globalsns, builtinsns) < 0)
+    {
+        return -1;
+    }
+    // We may consider relaxing these if it becomes a problem.
+    errmsg = _PyCode_CheckNoInternalState(co);
+    if (errmsg != NULL) {
```

Comment:
The convention is NULL means error and non-null is not error. I don't think it's a good idea to do the opposite here.

Reply:
The function cannot fail, but I see what you are saying.  I'll change this.

Reply:
fixed